datafusion/core/src/physical_planner.rs

```diff
@@ -156,13 +156,18 @@ fn create_physical_name(e: &Expr, is_first_expr: bool) -> Result<String> {
         Expr::Case(case) => {
             let mut name = "CASE ".to_string();
             if let Some(e) = &case.expr {
-                let _ = write!(name, "{e} ");
+                let _ = write!(name, "{} ", create_physical_name(e, false)?);
             }
             for (w, t) in &case.when_then_expr {
-                let _ = write!(name, "WHEN {w} THEN {t} ");
+                let _ = write!(
```

Comment:
I think this code makes `CASE` name generation consistent with name generation for other expressions (aka recursively call `create_physical_name` rather than using display impl)

So I don't understand why this is a "temporary" fix (in the sense of what is the longer term fix 🤔 ?)

> I think this requirement might be unreasonable. Having multiple different types of names for expressions, while also ensuring they are the same, seems somewhat burdensome and error-prone.

I agree the use of display name sometimes error prone. But I don't have any great ideas for an alternative

Reply:
@alamb These queries require that [physical_name()](https://github.com/apache/datafusion/blob/e9e2951abb3c5a0fb3e5d5b8ed75db4ad380c07a/datafusion/core/src/physical_planner.rs#L130) and [display_name()](https://github.com/apache/datafusion/blob/e9e2951abb3c5a0fb3e5d5b8ed75db4ad380c07a/datafusion/expr/src/expr.rs#L992) produce identical results.

For example, if the physical_name is modified like [this](https://github.com/jonahgao/datafusion/commit/ab8b13133bcf6e6236effa6fab8d2a76f5548e3c), these queries will fail again.

I think we should avoid having this implicit requirement. This may deserve further investigation, and there might be better solutions for a fix.